BatchResolveLock doesn't handle stale pessimistic locks with invalidated primary and may break data consistency

## Bug Report

\* This is found by reviewing code and not yet confirmed by tests.

As way know, pessimistic transactions may switch primary during execution, and a stale pessimistic lock's primary may point to a key that is no longer the primary of the transaction.

It's quite tricky to handle this case when resolving locks. Historically, we've done several fixes trying to make it correct, such as https://github.com/pingcap/tidb/pull/14787 , https://github.com/pingcap/tidb/pull/21689 and many more. However, we still find some incorrectly handled corner cases recently. One is https://github.com/pingcap/tidb/issues/42937 , and here's another one.

Since a stale pessimistic lock may points to a key that's not the actual primary of that transaction, the status of the key pointed by the pessimistic lock may not indicate the real state (committed or rolled back) of that transaction. We used to optimize resolving lock by caching the result of `check_txn_status`, and then it's found to be problematic and fixed in https://github.com/pingcap/tidb/pull/21689 .

However, we found that `BatchResolveLocks` (which is used in GC) have a different way to misuse the wrong transaction status. It's passed an array of locks (which is usually collected by `ScanLock` RPC), then checks their primary's status, collects the results in a list, and finally send them to TiKV in one `ResolveLock` RPC. There's no special handling for pessimistic locks, and it may send incorrect transaction state to TiKV. If the transaction is committed and some of their keys are prewritten but not committed (secondaries of 2PC transactions, or any keys of async commit transactions), these key may be incorrectly rolled back, causing the transaction incompletely committed and breaks the data consistency.

Constructed the problem by a test on unistore, but the test passes on TiKV. The reason is unknown yet.

```
--- FAIL: TestIssue43243 (1.15s)
    testkit.go:134:
                Error Trace:    /home/tidb/misono/sync/tidb/testkit/testkit.go:134
                                                        /home/tidb/misono/sync/tidb/testkit/testkit.go:127
                                                        /home/tidb/misono/sync/tidb/tests/realtikvtest/pessimistictest/pessimistic_test.go:3445
                Error:          Received unexpected error:
                                [admin:8223]data inconsistency in table: t2, index: v, handle: 1, index-values:"handle: 1, values: [KindInt64 1 KindInt64 1]" != record-values:""
                                github.com/pingcap/errors.AddStack
                                        /home/tidb/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/errors.go:174
                                github.com/pingcap/errors.(*Error).GenWithStackByArgs
                                        /home/tidb/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/normalize.go:164
                                github.com/pingcap/tidb/util/logutil/consistency.(*Reporter).ReportAdminCheckInconsistent
                                        /home/tidb/misono/sync/tidb/util/logutil/consistency/reporter.go:293
                                github.com/pingcap/tidb/executor.(*tableWorker).compareData.func2
                                        /home/tidb/misono/sync/tidb/executor/distsql.go:1336
                                github.com/pingcap/tidb/kv.(*HandleMap).Range
                                        /home/tidb/misono/sync/tidb/kv/key.go:452
                                github.com/pingcap/tidb/executor.(*tableWorker).compareData
                                        /home/tidb/misono/sync/tidb/executor/distsql.go:1334
                                github.com/pingcap/tidb/executor.(*tableWorker).executeTask
                                        /home/tidb/misono/sync/tidb/executor/distsql.go:1419
                                github.com/pingcap/tidb/executor.(*tableWorker).pickAndExecTask
                                        /home/tidb/misono/sync/tidb/executor/distsql.go:1150
                                github.com/pingcap/tidb/executor.(*IndexLookUpExecutor).startTableWorker.func1
                                        /home/tidb/misono/sync/tidb/executor/distsql.go:748
                                runtime.goexit
                                        /usr/local/go/src/runtime/asm_amd64.s:1598
                Test:           TestIssue43243
                Messages:       sql:admin check table t2, [], error stack [admin:8223]data inconsistency in table: t2, index: v, handle: 1, index-values:"handle: 1, values: [KindInt64 1 KindInt64 1]" != record-values:""
                                github.com/pingcap/errors.AddStack
                                        /home/tidb/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/errors.go:174
                                github.com/pingcap/errors.(*Error).GenWithStackByArgs
                                        /home/tidb/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/normalize.go:164
                                github.com/pingcap/tidb/util/logutil/consistency.(*Reporter).ReportAdminCheckInconsistent
                                        /home/tidb/misono/sync/tidb/util/logutil/consistency/reporter.go:293
                                github.com/pingcap/tidb/executor.(*tableWorker).compareData.func2
                                        /home/tidb/misono/sync/tidb/executor/distsql.go:1336
                                github.com/pingcap/tidb/kv.(*HandleMap).Range
                                        /home/tidb/misono/sync/tidb/kv/key.go:452
                                github.com/pingcap/tidb/executor.(*tableWorker).compareData
                                        /home/tidb/misono/sync/tidb/executor/distsql.go:1334
                                github.com/pingcap/tidb/executor.(*tableWorker).executeTask
                                        /home/tidb/misono/sync/tidb/executor/distsql.go:1419
                                github.com/pingcap/tidb/executor.(*tableWorker).pickAndExecTask
                                        /home/tidb/misono/sync/tidb/executor/distsql.go:1150
                                github.com/pingcap/tidb/executor.(*IndexLookUpExecutor).startTableWorker.func1
                                        /home/tidb/misono/sync/tidb/executor/distsql.go:748
                                runtime.goexit
                                        /usr/local/go/src/runtime/asm_amd64.s:1598
FAIL
```

The test code:

https://github.com/pingcap/tidb/blob/c350ffd9fd6e831f3507879a0e94c24789463781/tests/realtikvtest/pessimistictest/pessimistic_test.go#L3391-L3447

--- a/go.mod
+++ b/go.mod
@@ -63,7 +63,7 @@ require (
 	github.com/spf13/pflag v1.0.5
 	github.com/stretchr/testify v1.7.2-0.20220504104629-106ec21d14df
 	github.com/tiancaiamao/appdash v0.0.0-20181126055449-889f96f722a2
-	github.com/tikv/client-go/v2 v2.0.1-0.20230329072435-fc18f677df02
+	github.com/tikv/client-go/v2 v2.0.1-0.20230630082913-bf42043d98b6
 	github.com/tikv/pd/client v0.0.0-20220307081149-841fa61e9710
 	github.com/twmb/murmur3 v1.1.3
 	github.com/uber/jaeger-client-go v2.22.1+incompatible

